---
title: "Class 11: Structural Bioinformatics"
author: "Yi Fu"
date: "5/7/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```




## The PDB database

The [PDB](http://www.rcsb.org) is the main repository for biomolecular structure data.

Here we examine the contents of the PDB:

> Q1: Download a CSV file from the PDB site (accessible from “Analyze” -> “PDB Statistics” > “by Experimental Method and Molecular Type”. Move this CSV file into your RStudio project and determine the percentage of structures solved by X-Ray and Electron Microscopy. From the website what proportion of structures are protein? Aim to have a rendered GitHub document with working code that yields your answers.

```{r}
db = read.csv("Data Export Summary.csv", row.names = 1)
head(db)
```

We could also try the datapasta package and copy from the website and "Addins" > "Paste as data.frame"
```{r}
# if (!require("datapasta")) {
#   install.packages("datapasta")
# }
library("datapasta")
```
```{r}
## Use copy and paste
# data.frame(stringsAsFactors=FALSE,
#    Experimental.Method = c("X-Ray", "NMR", "Electron Microscopy", "Other",
#                            "Multi Method", "Total"),
#               Proteins = c(126880, 11062, 2277, 256, 129, 140604),
#          Nucleic.Acids = c(2012, 1279, 31, 4, 5, 3331),
#     Protein/NA.Complex = c(6547, 259, 800, 6, 2, 7614),
#                  Other = c(8, 8, 0, 13, 1, 30),
#                  Total = c(135447, 12608, 3108, 279, 137, 151579)
# )

```

How many are X=Ray, etc...
```{r}
(db$Total/sum(db$Total))*100
```

what proportion of structures are protein?
```{r}
sum(db$Proteins)/sum(db$Total)*100
```

> Q2: Type HIV in the PDB website search box on the home page and determine how many
HIV-1 protease structures are in the current PDB?

There are 1157 as of 2019-05-07 See:
http://www.rcsb.org/pdb/results/results.do?tabtoshow=Current&qrid=4C97A937


Reading PDB file data into R
```{r}
if (!require("bio3d")) {
  install.packages("bio3d")
}
library(bio3d)
```

```{r}
pdb = read.pdb("1hsg.pdb")
pdb
```

> Q6. How many amino acid residues are there in this pdb object and what are the two nonprotein residues?
#### protein residues: 1198. nonprotein residues: HOH (127), MK1 (1).

```{r}
attributes(pdb)
head(pdb$atom)
pdb$atom[1:2, c("eleno", "elety", "x","y","z")]
pdb$atom$elety[1:2]
plot.bio3d(pdb$atom$b[pdb$calpha], sse=pdb, typ="l", ylab="B-factor")
```

> Q7. What type of R object is pdb$atom? HINT: You can always use the str() function to get a useful summery of any R object.

```{r}
str(pdb$atom)
```

```{r}
pdb$xyz
dim(pdb$xyz)
pdb$xyz[ 1, atom2xyz(1:2) ]
```

Atom selection is done via the function **atom.select**

```{r}
inds = atom.select(pdb,"protein")
pdb$atom[inds$atom,]
inds = atom.select(pdb,"ligand")
pdb$atom[inds$atom,]
```

```{r}
prot.pdb = atom.select (pdb,"protein",value = T)
write.pdb (prot.pdb, file="1hsg_protein.pdb")

lig.pdb = atom.select (pdb,"ligand",value = T)
write.pdb (lig.pdb, file="1hsg_ligand.pdb")
```

## Search and retrieve Adenylate kinase structures

```{r}
aa <- get.seq("1ake_A")
b <- blast.pdb(aa)
hits <- plot(b)
head(hits$pdb.id)
```

```{r}
# Fetch PDBs
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)

# Align structures
pdbs <- pdbaln(files)

# Vector containing PDB codes
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels=ids)
```

```{r}
# Calculate sequence conservation
cons <- conserv(pdbs, method="entropy22")

# SSE annotations
sse <- pdbs2sse(pdbs, ind=1, rm.gaps=FALSE)

# Plot conservation per residue
plotb3(cons, sse=sse, ylab="Sequence entropy")
```

```{r}
anno <- pdb.annotate(ids)
print(unique(anno$source))
```

```{r}
# find invariant core
core <- core.find(pdbs)
# superimpose all structures to core
pdbs$xyz = pdbfit(pdbs, core)
# Perform PCA
pc.xray <- pca(pdbs)
# Calculate RMSD
rd <- rmsd(pdbs)
# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)
plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

